File: app/src/main/java/closer/vlllage/com/closer/MapsActivity.java
package closer.vlllage.com.closer;

import android.app.Fragment;
import android.app.FragmentManager;
import android.content.Intent;
import android.os.Bundle;

import com.github.queatz.slidescreen.SlideScreen;
import com.github.queatz.slidescreen.SlideScreenAdapter;

import closer.vlllage.com.closer.handler.PersonalSlideFragment;
import closer.vlllage.com.closer.handler.map.MapViewHandler;
import closer.vlllage.com.closer.handler.settings.SettingsSlideFragment;
import closer.vlllage.com.closer.pool.PoolActivity;

public class MapsActivity extends PoolActivity {

    public static final String EXTRA_LAT_LNG = "latLng";
    public static final String EXTRA_SUGGESTION = "suggestion";
    public static final String EXTRA_EVENT_ID = "eventId";
    public static final String EXTRA_NAME = "name";
    public static final String EXTRA_STATUS = "status";
    public static final String EXTRA_PHONE = "phone";
    public static final String EXTRA_MESSAGE = "message";
    public static final String EXTRA_SCREEN = "screen";
    public static final String EXTRA_SCREEN_MAP = "screen.map";
    public static final String EXTRA_SCREEN_PERSONAL = "screen.personal";
    public static final String EXTRA_SCREEN_SETTINGS = "screen.settings";

    public static final int POSITION_SCREEN_PERSONAL = 0;
    public static final int POSITION_SCREEN_MAP = 1;
    public static final int POSITION_SCREEN_SETTINGS = 2;
    public static final int NUM_SCREENS = 3;

    private SlideScreen slideScreen;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.slidescreen);
        slideScreen = findViewById(R.id.slideScreen);
        slideScreen.setAdapter(new SlideScreenAdapter() {
            @Override
            public int getCount() {
                return NUM_SCREENS;
            }

            @Override
            public Fragment getSlide(int position) {
                switch (position) {
                    case POSITION_SCREEN_PERSONAL: return new PersonalSlideFragment();
                    case POSITION_SCREEN_MAP: return $(MapViewHandler.class).getMapFragment();
                    case POSITION_SCREEN_SETTINGS: return new SettingsSlideFragment();
                    default: throw new IndexOutOfBoundsException();
                }
            }

            @Override
            public FragmentManager getFragmentManager() {
                return MapsActivity.this.getFragmentManager();
            }
        });

        if (getIntent() != null) {
            onNewIntent(getIntent());
        }

        $(MapViewHandler.class).setOnRequestMapOnScreenListener(() -> {
            slideScreen.setSlide(POSITION_SCREEN_MAP);
        });

        slideScreen.setSlide(POSITION_SCREEN_MAP);
    }

    @Override
    protected void onNewIntent(Intent intent) {
        if (intent.hasExtra(EXTRA_SCREEN)) {
            switch (intent.getStringExtra(EXTRA_SCREEN)) {
                case EXTRA_SCREEN_PERSONAL:
                    slideScreen.setSlide(POSITION_SCREEN_PERSONAL);
                    return;
                case EXTRA_SCREEN_MAP:
                    slideScreen.setSlide(POSITION_SCREEN_MAP);
                    return;
                case EXTRA_SCREEN_SETTINGS:
                    slideScreen.setSlide(POSITION_SCREEN_SETTINGS);
                    return;
            }
        }

        $(MapViewHandler.class).handleIntent(intent);
    }

    @Override
    public void onBackPressed() {
        $(MapViewHandler.class).onBackPressed(result -> {
            if (result) {
                return;
            }

            if (!slideScreen.isExpose()) {
                slideScreen.expose(true);
                return;
            }

            super.onBackPressed();
        });
    }
}

File: app/src/main/java/closer/vlllage/com/closer/MapsActivity.kt
package closer.vlllage.com.closer

import android.content.Intent
import android.os.Bundle
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentContainerView
import closer.vlllage.com.closer.databinding.MainBinding
import closer.vlllage.com.closer.handler.data.PersistenceHandler
import closer.vlllage.com.closer.handler.helpers.DisposableHandler
import closer.vlllage.com.closer.handler.helpers.ScanQrCodeHandler
import closer.vlllage.com.closer.handler.helpers.ToastHandler
import closer.vlllage.com.closer.handler.map.MapViewHandler
import closer.vlllage.com.closer.handler.welcome.WelcomeSlideFragment
import closer.vlllage.com.closer.pool.PoolActivity
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers

class MapsActivity : PoolActivity() {

    private lateinit var fragmentContainer: FragmentContainerView

    private val accessDisposableGroup = on<DisposableHandler>().group()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        fragmentContainer = MainBinding.inflate(layoutInflater).let {
            setContentView(it.root)
            it.fragmentContainer
        }

        refreshScreens()

        if (intent != null) {
            onNewIntent(intent)
        }

        if (on<PersistenceHandler>().access.not()) {
            on<PersistenceHandler>().changes
                    .filter { it == PersistenceHandler.PREFERENCE_ACCESS }
                    .map { on<PersistenceHandler>().access }
                    .filter { it }
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe {
                        refreshScreens()
                        accessDisposableGroup.dispose()
                    }.also {
                        accessDisposableGroup.add(it)
                    }
        }
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)

        if (intent.action == Intent.ACTION_VIEW) {
            intent.data?.let {
                if (it.host == "closer.group") on<ScanQrCodeHandler>().handleResult(it)
            }
        }

        if (intent.hasExtra(EXTRA_PROMPT)) {
            on<ToastHandler>().show(intent.getStringExtra(EXTRA_PROMPT)!!)
        }

        on<MapViewHandler>().handleIntent(intent)
    }

    override fun onBackPressed() {
        if (on<PersistenceHandler>().access) {
            on<MapViewHandler>().onBackPressed {
                if (it) {
                    return@onBackPressed
                }

                super.onBackPressed()
            }
        } else {
            super.onBackPressed()
        }
    }

    private fun refreshScreens() {
        if (on<PersistenceHandler>().access.not()) {
            show(WelcomeSlideFragment())
        } else {
            show(on<MapViewHandler>().mapFragment)
        }
    }

    private fun show(fragment: Fragment) {
        supportFragmentManager.beginTransaction()
            .replace(R.id.fragmentContainer, fragment)
            .commitAllowingStateLoss()
    }

    companion object {
        const val EXTRA_LAT_LNG = "latLng"
        const val EXTRA_ZOOM = "latLngZoom"
        const val EXTRA_SUGGESTION = "suggestion"
        const val EXTRA_EVENT_ID = "eventId"
        const val EXTRA_PHONE = "phone"
        const val EXTRA_PROMPT = "prompt"
    }
}
